src/backend/distributed/executor/subplan_execution.c

```diff
@@ -66,13 +64,35 @@ ExecuteSubPlans(DistributedPlan *distributedPlan)
 		DestReceiver *copyDest = NULL;
 		ParamListInfo params = NULL;
 		EState *estate = NULL;
-
+		bool writeLocalFile = false;
 		char *resultId = GenerateResultId(planId, subPlanId);
+		List *workerNodeList =
+			FindAllWorkerNodesUsingSubplan(intermediateResultsHash, resultId);
+
+		/*
+		 * Write intermediate results to local file only if there is no worker
+		 * node that receives them.
+		 *
+		 * This could happen in two cases:
+		 * (a) Subquery in the having
+		 * (b) The intermediate result is not used, such as RETURNING of a
+		 *     modifying CTE is not used
+		 *
+		 * For SELECT, Postgres/Citus is clever enough to not execute the CTE
+		 * if it is not used at all, but for modifications we have to execute
+		 * the queries.
+		 */
+		if (workerNodeList == NIL)
+		{
+			writeLocalFile = true;
+			elog(DEBUG1, "%s is not broadcasted to any node, write to local only",
```

Comment:
At this point it is not written to a local file. I think in the future this might make debugging harder because we will see this debugging message, thinking that it is written to a local file, but maybe there is a failure so it is not written to a local file. Maybe we can have something like `will write to only local`

Reply:
Agreed, this should be rephrased, or maybe just omitted entirely.

Reply:
This debug message is really helpful in the regression tests. I am more inclined to rephrase it. How does `Subplan %s will not broadcasted to any node, will write to local only` sound?

Reply:
Updated all my `DEBUG1` messages. Now they are in two different forms:
```
DEBUG:  Subplan 18_1 will not broadcasted to any node, will write to local only
DEBUG:  Subplan 18_2 will be broadcasted to localhost:57637
```

Reply:
It should be `will not be broadcasted`.

Reply:
```diff
-DEBUG:  Subplan 184_1 will not broadcasted to any node, will write to local only
+DEBUG:  Subplan 184_1 will not be broadcasted to any node, will be written to local only
```

Reply:
I hid the message behind a GUC `log_intermediate_results`